### PR TITLE
ESR68 migration guide

### PR DESCRIPTION
**Pull request by ptomato**

### CODE REVIEW

File: docs/Building SpiderMonkey.md [hunk 1]
```diff
@@ -12,19 +12,23 @@ also usually available through a package manager.
...
 ## Getting the source code ##
 
 Currently, the most reliable way to get the SpiderMonkey source code is
 to download the Firefox source.
-At the time of writing, the latest source for Firefox ESR 60, which
-contains the source for SpiderMonkey ESR 60, can be found here:
-http://ftp.mozilla.org/pub/firefox/releases/60.3.0esr/source/
+At the time of writing, the latest source for Firefox ESR 68, which
+contains the source for SpiderMonkey ESR 68, can be found here:
+http://ftp.mozilla.org/pub/firefox/releases/FIXME
```
Comment (moztcampbell) [inline docs/Building SpiderMonkey.md]: https://ftp.mozilla.org/pub/firefox/releases/68.1.0esr/source/

File: docs/Migration Guide.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+It is no longer necessary to initialize `JS::GCHashMap<>` with `init()`.
+
+**Recommendation:** Simply remove the call to `init()`.
+
+### Principals ###
+
+Functions having to do with stack frames, such as
+`JS::GetSavedFrameSource()` and others, now take a `JSPrincipals*`
+pointer as an argument.
+**FIXME**: What is that for?
```
Comment (moztcampbell) [inline docs/Migration Guide.md]: JSPrincipals is a trust token the indicates the privileges of the current caller. The stack frame APIs use this to appropriately hide system stack frames from less trusted code.

File: docs/Migration Guide.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+**Recommendations:**
+- **FIXME**
```
Comment (moztcampbell) [inline docs/Migration Guide.md]: If you are not doing sophisticated things with principals, you can find the current context's principal with `cx->realm()->principals();`

File: docs/Migration Guide.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+**Recommendation**: Remove all `JSAutoRequest`, `JS_BeginRequest()`, and
+`JS_EndRequest()` from your code.
+
+### Rename compartments to realms ###
+
+In ESR 68, a "realm" is the term for the environment belonging to a
+particular global object, so when you enter the "realm" of a global
+object you are executing code with that object as the global object.
+As before, with compartments, objects from different realms cannot
+access each other unless they are "wrapped".
```
Comment (hotsphink) [inline docs/Migration Guide.md]: Same-compartment different-realm objects do *not* need to be wrapped. The first sentence is still correct.

Objects in different compartments still need be wrapped. You will want to enter the appropriate realm, and there's less checking if you're doing it the right way when accessing objects from different realms (unless you run with the equivalent of the JS shell's --more-compartments, which I think does compartment-per-realm.)

File: docs/Migration Guide.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+Some APIs that previously operated explicitly on a global object, now
+operate on the current realm's global object, but there are currently no
+examples of that in this code.
+
+"Compartments" still exist, but they are now more focused on security,
+and are used for isolating code in different browser tabs in Firefox,
+for example.
+Realms are used within the same compartment for debugger code vs. client
+code, for example.
+All code in these examples is run within the same compartment.
```
Comment (hotsphink) [inline docs/Migration Guide.md]: Does any of it run in different realms? If not, then this should say "...is run within the same realm."

File: docs/Migration Guide.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  There is an appropriate overload to use instead.
+- `JS_GetGlobalForObject()` → `JS::GetNonCCWObjectGlobal()`
+- `JS_NewArrayBuffer()` → `JS::NewArrayBuffer()`; ditto for
+  `JS_NewArrayBufferWithContents()` and `JS_NewExternalArrayBuffer()`.
+  In addition, `JS::NewExternalArrayBuffer()` has dropped support for
+  the ref callback.
+- `JS_ObjectIsFunction()` → `js::IsFunctionObject()`
+- `JS_ReportWarningASCII()` → `JS::WarnASCII()`
+- `JS_ReportWarningUTF8()` → `JS::WarnUTF8()`
+- `JS_updateMallocCounter()` → ??? **FIXME**
```
Comment (hotsphink) [inline docs/Migration Guide.md]: Looks like it's JS::AddAssociatedMemory now.

File: docs/Migration Guide.md (hunk 2 above)
Comment (ptomato) [reply docs/Migration Guide.md]: Included verbatim in the docs now, thanks :smile:

File: docs/Migration Guide.md (hunk 6 above)
Comment (ptomato) [reply docs/Migration Guide.md]: I couldn't really figure out how to use that without having it crash; in any case it seems not to be a drop-in replacement. (And could we add `JS::MemoryUse::Embedding`?)

**Comment by ptomato** [comment]
Merging, and setting the esr68 branch as the main branch, after IRC discussion today.

**Pull request merged**